fix(remix-react): fix `<Form>` submit to not break `formMethod` functionality

nrako opened the pull request:
<!--

👋 Hey, thanks for your interest in contributing to Remix!

If this is a simple docs change, go ahead and delete all this.

**Please ask first before starting work on any significant new features.**

It's never a fun experience to have your pull request declined after investing a
lot of time and effort into a new feature. To avoid this from happening, we
request that contributors create a
[Feature Request discussion](https://github.com/remix-run/remix/discussions/new?category=ideas)
to first discuss any significant new features.

https://github.com/remix-run/remix/blob/main/CONTRIBUTING.md

Please fill in or delete each item below:

-->

Closes: #2162 (supersede) – credits to @mdoury for his original PR

> the `<Form>` component – unlike native `<form>` - does not respect the `formmethod` attribute set
on the submitter (the button submitting the form).

- [x] ~~Docs~~ N/A
- [x] Tests

Testing Strategy:
```jsx
<Form method="post">
  <input name="eat" value="pizza" />
  <button type="submit"> Submit with GET </button>
</Form>
// should submit with a GET to `/?eat=pizza` but instead do a POST
```

See integration [Bug-Report-Test](https://github.com/remix-run/remix/blob/37b0140f80dfb333240da96fb9ff83e713848458/integration/bug-report-test.ts) via commit: https://github.com/remix-run/remix/commit/37b0140f80dfb333240da96fb9ff83e713848458

---

**test: failing test with `<Form>` not respecting `formMethod`**
06333c1ed &lt;Nicholas Rakoto&gt; Thu, 30 Jun 2022 00:33:15 +0200

    Bug report integration test demonstrating how the `<Form>` component – 
    unlike native `<form>` - does not respect the `formmethod` attribute set on
    the submitter (the `<button>` submitting the form).


**fix(remix-react): submit `<Form>` w/method set by submitter `formmethod`**
bfd1f88de &lt;Nicholas Rakoto&gt; Thu, 30 Jun 2022 00:32:22 +0200

    Fix `<Form>` component by not passing the form's `method` into the
    `options` when calling the `submit(target, options)` function (created with
    the `useSubmit()` hook). This ensures that the option does not take 
    precedence on any `formethod` set on the submitter which otherwise break the
    `formmethod` functionality.

    Since #3094 and #3094 the `<Form>` passes the "submitter" (if any) as the
    `target` to the `submit(target, options)` (`useSubmitImpl`) which will
    attempt to read the `formmethod`, `formaction` and `formenctype` attributes
    on the target. Therefore, the responsability to infer which method should be
    used should solely be on `useSubmitImpl` for the given
    `target`.

    Co-authored-by: Maxime Doury <maxime.doury@docavenue.com>

MichaelDeBoey reviewed (changes requested):
Could you maybe add the other tests from #2162 as well please?

nrako commented:
@MichaelDeBoey – could you a bit more explicit about the "other tests". The tests in this PR specifically cover what is fixed, which is to ensure that the `formethod` on a submit button is not ignored. The only other test I see in the PR is one that more or less verify that a `<Form>` submit with the `method` specified which I believe is already covered by many other existing tests.

Note: I've refactored the tests a bit, fixed the failing test, but the tests under a new `describe` and split the assertions in two different tests for both POST and GET.

evanstern commented:
Any movement on this? Is there any way I can help get this across the finish line?

MichaelDeBoey commented:
@nrako The tests in #2162 have split tests for using the `form` `method` attribute, the `button` `formmethod` attribute & combined, you only have tests for the combined case

MichaelDeBoey commented:
@nrako The tests in #2162 have split tests for using the `form` `method` attribute, the `button` `formmethod` attribute & combined, you only have tests for the combined case

evanstern commented:
If someone can give me some instructions on how I can make these changes, I'd be happy to make them so this can get merged. This is functionality that I desperately need for my project.

kevlened commented:
I opened a PR at #4053 that adds the tests from #2162

MichaelDeBoey commented:
Closing this one in favor of #4053
